releasing/plugin-commands-deploy/src/deployCatalogHook.ts

```diff
@@ -0,0 +1,51 @@
+import { matchCatalogResolveResult, resolveFromCatalog } from '@pnpm/catalogs.resolver'
+import { type Catalogs } from '@pnpm/catalogs.types'
+import { type ProjectManifest, DEPENDENCIES_FIELDS } from '@pnpm/types'
+
+/**
+ * Teach the `pnpm deploy` command how to interpret the catalog: protocol.
+ *
+ * This is a hack to work around a design problem between pnpm deploy and
+ * catalogs.
+ *
+ *   - The catalog protocol is intentionally only allowed to be used by
+ *     importers. External dependencies cannot use the catalog: protocol by
+ *     design.
+ *   - When using pnpm deploy, dependency workspace packages aren't considered
+ *     "importers".
+ *
+ * To work around the conflict of designs above, this readPackage hook exists to
+ * make catalogs usable by non-importers specifically on pnpm deploy.
+ *
+ * Unfortunately this introduces a correctness issue where the catalog: protocol
+ * is replaced for all packages (even external dependencies), not just packages
```

Comment:
Here's a commit testing that out: https://github.com/pnpm/pnpm/commit/273d8035748836ef136ce587fa5c89a6e5f0d5a4

I think the tests for that change are failing because there's 2 places we call `readPackageHook`, and one of them doesn't pass the `dir` argument.

https://github.com/pnpm/pnpm/blob/f30590bdfc9638888355f8ce82a3bfc4e812e13c/pkg-manager/resolve-dependencies/src/resolveDependencies.ts#L1319

https://github.com/pnpm/pnpm/blob/0406d4ad3d2a10326cd46ccc242aadad878a5ead/pkg-manager/get-context/src/index.ts#L167

If I run,

```
pnpm --filter=./pkg-manager/core deploy ./target
```

The workspace packages are installed using the former case that doesn't get a `dir` argument.
<img width="1397" alt="Screenshot 2024-07-12 at 5 34 34 PM" src="https://github.com/user-attachments/assets/c4898711-c6e0-4c41-b2c6-b04401138704">

I can investigate a bit more, but I think this is because `dependenciesMeta[*].injected` workspace packages aren't treated as importers.